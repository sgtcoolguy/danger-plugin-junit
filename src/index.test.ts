import junit from "./index"

declare const global: any

describe("junit()", () => {
  beforeEach(() => {
    global.warn = jest.fn()
    global.message = jest.fn()
    global.fail = jest.fn()
    global.markdown = jest.fn()
  })

  afterEach(() => {
    global.warn = undefined
    global.message = undefined
    global.fail = undefined
    global.markdown = undefined
  })

  it("Needs a real xml file", async () => {
    await junit({})

    expect(global.fail).not.toHaveBeenCalled()
    expect(global.markdown).not.toHaveBeenCalled()
    expect(global.warn).toHaveBeenCalledTimes(1)
    expect(global.warn).toHaveBeenCalledWith(
      ":mag: Can't find junit reports at `./build/reports/**/TESTS*.xml`, skipping generating JUnit Report.",
    )
  })

  it("Gives success summary when no failures", async () => {
    await junit({
      pathToReport: "./fixtures/junit_success.xml",
    })

    expect(global.message).toHaveBeenCalledTimes(1)
    expect(global.message).toHaveBeenCalledWith(
      ":white_check_mark: All tests are passing\nNice one! All 3 tests are passing.",
    )
    expect(global.fail).toHaveBeenCalledTimes(0)
    expect(global.markdown).toHaveBeenCalledTimes(0)
  })

  it("Checks test failures were failed", async () => {
    await junit({
      pathToReport: "./fixtures/junit_failures.xml",
    })

    expect(global.message).toHaveBeenCalledTimes(1)
    expect(global.message).toHaveBeenCalledWith(
      ":x: 1 tests have failed\nThere are 1 tests failing and 3 skipped out of 19 total tests.",
    )
    expect(global.fail).toHaveBeenCalledTimes(1)
    expect(global.fail).toHaveBeenCalledWith("Tests have failed, see below for more information.")
    expect(global.markdown).toHaveBeenCalledTimes(1)
    expect(global.markdown.mock.calls[0][0]).toMatchSnapshot()
  })

  it("Combines multiple report files", async () => {
    await junit({
      pathToReport: "./fixtures/junit_*.xml",
    })

    expect(global.message).toHaveBeenCalledTimes(1)
    expect(global.message).toHaveBeenCalledWith(
      ":x: 2 tests have failed\nThere are 2 tests failing and 3 skipped out of 26 total tests.",
    )
    expect(global.fail).toHaveBeenCalledTimes(1)
    expect(global.fail).toHaveBeenCalledWith("Tests have failed, see below for more information.")
    expect(global.markdown).toHaveBeenCalledTimes(1)
    expect(global.markdown.mock.calls[0][0]).toMatchSnapshot()
  })

  it("handles karma junit report", async () => {
    await junit({
      pathToReport: "./fixtures/TESTS-android.xml",
    })

    expect(global.message).toHaveBeenCalledTimes(1)
    expect(global.message).toHaveBeenCalledWith(
      ":x: 1 tests have failed\nThere are 1 tests failing and 1 skipped out of 2 total tests.",
    )
    expect(global.fail).toHaveBeenCalledTimes(1)
    expect(global.fail).toHaveBeenCalledWith("Tests have failed, see below for more information.")
    expect(global.markdown).toHaveBeenCalledTimes(1)
    expect(global.markdown.mock.calls[0][0]).toMatchSnapshot()
  })

  it("Handles failures from mocha-jenkins-reporter", async () => {
    await junit({
      pathToReport: "./fixtures/junit_report.xml",
    })

    expect(global.message).toHaveBeenCalledTimes(1)
    expect(global.message).toHaveBeenCalledWith(
      ":x: 1 tests have failed\nThere are 1 tests failing and 0 skipped out of 4 total tests.",
    )
    expect(global.fail).toHaveBeenCalledTimes(1)
    expect(global.fail).toHaveBeenCalledWith("Tests have failed, see below for more information.")
    expect(global.markdown).toHaveBeenCalledTimes(1)
    expect(global.markdown.mock.calls[0][0]).toMatchSnapshot()
  })

  it("Handles skipped from karma-titanium-launcher", async () => {
    await junit({
      pathToReport: "./fixtures/TESTS-ios.xml",
    })

    expect(global.message).toHaveBeenCalledTimes(1)
    expect(global.message).toHaveBeenCalledWith(
      `:white_check_mark: All tests are passing
Nice one! All 178 tests are passing.
(There are 8 skipped tests not included in that total)`,
    )
    expect(global.fail).toHaveBeenCalledTimes(0)
    expect(global.markdown).toHaveBeenCalledTimes(0)
  })

  it("allows setting a custom header", async () => {
    await junit({
      pathToReport: "./fixtures/junit_failures.xml",
      name: "My awesome tests",
    })

    expect(global.message).toHaveBeenCalledWith(
      ":x: 1 tests have failed\nThere are 1 tests failing and 3 skipped out of 19 total tests.",
    )
    expect(global.fail).toHaveBeenCalledTimes(1)
    expect(global.fail).toHaveBeenCalledWith("My awesome tests have failed, see below for more information.")
    expect(global.markdown).toHaveBeenCalledTimes(1)
    expect(global.markdown.mock.calls[0][0]).toMatchSnapshot()
  })

  it("allows reporting failures as warnings not errors", async () => {
    await junit({
      pathToReport: "./fixtures/junit_failures.xml",
      onlyWarn: true
    })

    expect(global.message).toHaveBeenCalledWith(
      ":x: 1 tests have failed\nThere are 1 tests failing and 3 skipped out of 19 total tests.",
    )
    expect(global.fail).toHaveBeenCalledTimes(0)
    expect(global.warn).toHaveBeenCalledTimes(1)
    expect(global.warn).toHaveBeenCalledWith("Tests have failed, see below for more information.")
    expect(global.markdown).toHaveBeenCalledTimes(1)
    expect(global.markdown.mock.calls[0][0]).toMatchSnapshot()
  })
})
