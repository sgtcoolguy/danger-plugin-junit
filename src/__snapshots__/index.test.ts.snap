// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`junit() Checks test failures were failed 1`] = `
"### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.Titanium.UI.Window</td><td>.safeAreaPadding with extendSafeArea true</td><td>0.052</td><td><pre>AssertionError: expected 0 to be above 0
at Assertion.prop.(anonymous function) (/should.js:1411:20)
		at Window.&lt;anonymous&gt; (/ti.ui.window.test.js:675:35)
		at Window.value (ti:/events.js:50:21)
		at Window.value (ti:/events.js:102:19)</pre></td></tr>
</table>
"
`;

exports[`junit() Combines multiple report files 1`] = `
"### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.Titanium.UI.Window</td><td>.safeAreaPadding with extendSafeArea true</td><td>0.052</td><td><pre>AssertionError: expected 0 to be above 0
at Assertion.prop.(anonymous function) (/should.js:1411:20)
		at Window.&lt;anonymous&gt; (/ti.ui.window.test.js:675:35)
		at Window.value (ti:/events.js:50:21)
		at Window.value (ti:/events.js:102:19)</pre></td></tr>
<tr><td>adb with an emulator running</td><td>#pull()</td><td>0</td><td><pre>ENOENT: no such file or directory, unlink '/Users/eharris/Documents/git/node-titanium-sdk/tests/build.prop'</pre></td></tr>
</table>
"
`;

exports[`junit() Handles failures from mocha-jenkins-reporter 1`] = `
"### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>adb with an emulator running</td><td>#pull()</td><td>0</td><td><pre>ENOENT: no such file or directory, unlink '/Users/eharris/Documents/git/node-titanium-sdk/tests/build.prop'</pre></td></tr>
</table>
"
`;

exports[`junit() allows setting a custom header 1`] = `
"### My awesome tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.Titanium.UI.Window</td><td>.safeAreaPadding with extendSafeArea true</td><td>0.052</td><td><pre>AssertionError: expected 0 to be above 0
at Assertion.prop.(anonymous function) (/should.js:1411:20)
		at Window.&lt;anonymous&gt; (/ti.ui.window.test.js:675:35)
		at Window.value (ti:/events.js:50:21)
		at Window.value (ti:/events.js:102:19)</pre></td></tr>
</table>
"
`;

exports[`junit() handles karma junit report 1`] = `
"### Tests:

<table><tr><th>Name</th><th>Time</th><th>Classname</th><th>Error</th></tr>
<tr><td>appcelerator.encrypteddatabase can be required</td><td>0.014</td><td>android.appcelerator.encrypteddatabase</td><td><pre>
TypeError: Cannot set property 'invocationAPIs' of undefined
error properties: Object({ toJSON: Function })
    at &lt;Jasmine&gt;
    at Object.moduleBootstrap [as bootstrap] (appcelerator.encrypteddatabootstrap.js:20:45)
    at Module.loadExternalModule (ti:/module.js:201:31)
    at Module.loadCoreModule (ti:/module.js:331:16)
    at Module.require (ti:/module.js:246:16)
    at require (ti:/module.js:570:15)
    at &lt;Jasmine&gt;</pre></td></tr>
</table>
"
`;
